File: backend/services/scraper.js
import puppeteer from 'puppeteer';

export async function scrapeGoogleReviews(query, location) {
  const browser = await puppeteer.launch({ headless: "new" });
  const page = await browser.newPage();
  
  try {
    // Google Maps search URL
    await page.goto(`https://www.google.com/maps/search/${encodeURIComponent(query)}+${encodeURIComponent(location)}`);
    
    // Wait for results to load
    await page.waitForSelector('.section-result');
    
    // Extract data
    const results = await page.evaluate(() => {
      const items = document.querySelectorAll('.section-result');
      return Array.from(items).map(item => ({
        name: item.querySelector('.section-result-title')?.textContent,
        rating: item.querySelector('.section-result-rating')?.textContent,
        reviews: item.querySelector('.section-result-num-reviews')?.textContent,
        address: item.querySelector('.section-result-location')?.textContent,
      }));
    });
    
    return results;
  } finally {
    await browser.close();
  }
}

export async function scrapeTripAdvisor(query, location) {
  const browser = await puppeteer.launch({ headless: "new" });
  const page = await browser.newPage();
  
  try {
    // TripAdvisor search URL
    await page.goto(`https://www.tripadvisor.com/Search?q=${encodeURIComponent(query)}+${encodeURIComponent(location)}`);
    
    // Wait for results to load
    await page.waitForSelector('.result-title');
    
    // Extract data
    const results = await page.evaluate(() => {
      const items = document.querySelectorAll('.result-item');
      return Array.from(items).map(item => ({
        name: item.querySelector('.result-title')?.textContent,
        rating: item.querySelector('.ui_bubble_rating')?.getAttribute('alt'),
        reviews: item.querySelector('.review-count')?.textContent,
        description: item.querySelector('.description')?.textContent,
      }));
    });
    
    return results;
  } finally {
    await browser.close();
  }
}

File: python/scraper.py
import requests
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import json
import time

class Scraper:
    def __init__(self):
        chrome_options = Options()
        chrome_options.add_argument("--headless")
        chrome_options.add_argument("--no-sandbox")
        chrome_options.add_argument("--disable-dev-shm-usage")
        self.driver = webdriver.Chrome(options=chrome_options)

    def scrape_google_reviews(self, query, location):
        try:
            search_url = f"https://www.google.com/maps/search/{query}+{location}"
            self.driver.get(search_url)
            
            # Wait for results to load
            WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CLASS_NAME, "section-result"))
            )
            
            # Click to expand reviews
            reviews_button = self.driver.find_element(By.CSS_SELECTOR, "[aria-label='Reviews']")
            reviews_button.click()
            time.sleep(2)  # Wait for reviews to load
            
            # Extract reviews
            reviews = []
            review_elements = self.driver.find_elements(By.CLASS_NAME, "section-review-content")
            
            for element in review_elements[:10]:  # Get first 10 reviews
                review = {
                    'text': element.find_element(By.CLASS_NAME, "section-review-text").text,
                    'rating': element.find_element(By.CLASS_NAME, "section-review-stars").get_attribute("aria-label"),
                    'date': element.find_element(By.CLASS_NAME, "section-review-date").text
                }
                reviews.append(review)
            
            return reviews
            
        except Exception as e:
            print(f"Error scraping Google reviews: {str(e)}")
            return []

    def scrape_tripadvisor(self, query, location):
        try:
            search_url = f"https://www.tripadvisor.com/Search?q={query}+{location}"
            self.driver.get(search_url)
            
            # Wait for results to load
            WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located((By.CLASS_NAME, "result-title"))
            )
            
            # Extract basic information
            results = []
            result_elements = self.driver.find_elements(By.CLASS_NAME, "result-item")
            
            for element in result_elements[:5]:  # Get first 5 results
                result = {
                    'name': element.find_element(By.CLASS_NAME, "result-title").text,
                    'rating': element.find_element(By.CLASS_NAME, "ui_bubble_rating").get_attribute("alt"),
                    'reviews': element.find_element(By.CLASS_NAME, "review-count").text,
                    'description': element.find_element(By.CLASS_NAME, "description").text
                }
                results.append(result)
            
            return results
            
        except Exception as e:
            print(f"Error scraping TripAdvisor: {str(e)}")
            return []

    def __del__(self):
        self.driver.quit()